### PR TITLE
Overhaul Dockerfile

### PR DESCRIPTION
**Pull request by theAkito**
* Use Alpine rather than Debian
* Remove unnecessary items
* Consolidate `RUN` directives
* Slimmify

### CODE REVIEW

**Comment by kevinastone** [comment]
Could you summarize the benefits of these changes?  Given this is mostly an educational project, I'd be hesitant to merge these changes unless they provide more clarity towards best practices for building Docker images for rust projects.

**Comment by theAkito** [comment]
### Benefits
* Using [Alpine](https://www.alpinelinux.org/) is always the number one choice for Docker images. [This](https://nickjanetakis.com/blog/the-3-biggest-wins-when-using-alpine-as-a-base-docker-image) is one of the tons of blog posts highlighting the biggest reasons, why Alpine should be used over any other OS inside a Docker container. Basically, it just gives benefits, without downsides, except when you truly need that `libc` environment, without any possibility to switch away from it.
* Using less `RUN` directives, less image layers are produced. Less image layers, means smaller image size & potentially quicker builds, especially if you strategically apply those directives, for example by applying a single one per logical unit in the Dockerfile. [This](https://stackoverflow.com/a/51660942/7061105) is one of the many explanations of how Docker image layers work.
* The `ENTRYPOINT` was kind of _point_less, if it's not under the last `FROM`' statement.
* Since we are dealing with Docker images resulting in containers here, there is no point in putting the binary into the path, where you would usually put it, like e.g. `/usr/local/bin`, because you are pointing to it directly, in the `ENTRYPOINT` either way. Reducing clutter like this improves readability & therefore maintainability of the Dockerfile.

**Comment by kevinastone** [comment]
Sorry about the slow response (got caught up at work this week).

I appreciate your points and the work here.  I think there's not a universal consensus for Alpine in the container space (or rust community).  The use of MUSL for the libc creates some compatibilities (though unlikely for this simplistic effort).  Because this is predominately a learning project, I don't think optimizing for the reduced image sizes are worth deviating from the stock rust image (which is Debian based).

I think the other changes might be beneficial if the goal was squeezing the last optimization, but I think the current readability is preferred.

Regarding `ENTRYPOINT` and putting the binary in the `/usr/local/bin`, I don't fully recall the decision, but it was likely better experiences when attaching to a running container or running the container with manual command arguments to trigger cargo commands.

Given all this, I'd prefer not to merge these changes.  Thanks for the effort.

**Comment by theAkito** [comment]
> I think there's not a universal consensus for Alpine in the container space (or rust community).

Docker is agnostic of what happens inside its containers.

There **is** a universal consensus, that Alpine is the base image to go. I have proven that to you with sources & there are tons of other sources online. Always use Alpine, except you have very strict & important reasons not to.



> I don't think optimizing for the reduced image sizes are worth deviating from the stock rust image (which is Debian based).

https://hub.docker.com/_/rust & CTRL+F for "alpine"....

There is no "stock" image. There are different variants & Alpine is always the default, except for strict legacy software, which relies too much on Debian based dependencies, etc.



> Regarding `ENTRYPOINT` and putting the binary in the `/usr/local/bin`, I don't fully recall the decision, but it was likely better experiences when attaching to a running container or running the container with manual command arguments to trigger cargo commands.

Attaching to a running container should only be done to debug an image, that is work in progress. In production, you shouldn't ever enter a container.

____

You seriously need to read up on how Docker works & what the whole point of it is, if this learning project also includes learning about Docker. You are misunderstanding tons about it. For example, Rust is not related to Docker, at all. It's just another thing, that runs inside a Docker container. Nothing more, nothing less. So, there won't be a consensus in the Rust community, as it will always be whatever the Docker standard is & that is Alpine.

**Alpine has the universal consensus to be the go-to base image.**
